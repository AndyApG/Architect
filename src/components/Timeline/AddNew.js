import React from 'react';
import PropTypes from 'prop-types';
import { Icon } from '../../ui/src/components';
import { Zoom } from '../../behaviours';
import { getCSSVariableAsString } from '../../utils/CSSVariables';

const zoomColors = [getCSSVariableAsString('--light-background'), getCSSVariableAsString('--light-background')];

const AddNewButton = Zoom(
  ({ onInsertStage }) => (
    <button
      className="timeline-add-new__button"
      onClick={onInsertStage}
      style={{ display: 'inline-block' }}
    >
      <div className="timeline-add-new__button-label">Add new stage here</div>
      <Icon name="add-a-screen" className="timeline-add-new__button-icon" />
    </button>
  ),
);

const TimelineAddNew = ({ onInsertStage }) => (
  <div className="timeline-add-new">
    <AddNewButton
      zoomColors={zoomColors}
      onInsertStage={onInsertStage}
    />
  </div>
);

TimelineAddNew.propTypes = {
  onInsertStage: PropTypes.func.isRequired,
};

export default TimelineAddNew;
