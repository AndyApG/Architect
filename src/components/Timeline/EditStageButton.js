import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';
import { motion } from 'framer-motion/dist/framer-motion';
import timelineImages from '@app/images/timeline';
import filterIcon from '@app/images/timeline/filter-icon.svg';
import skipLogicIcon from '@app/images/timeline/skip-logic-icon.svg';

const getTimelineImage = (type) => get(timelineImages, type, timelineImages.Default);

const EditStageButton = React.forwardRef(({
  hasFilter,
  hasSkipLogic,
  label,
  layoutId,
  onEditStage,
  type,
}, ref) => (
  <div
    className="timeline-stage__edit-stage"
    onClick={onEditStage}
  >
    <motion.div
      className="timeline-stage__screen"
      layoutId={layoutId}
      role="button"
      tabIndex="0"
      ref={ref}
    >
      <div className="timeline-stage__screen-preview">
        {
          getTimelineImage(type)
          && (
          <img
            src={getTimelineImage(type)}
            alt={`${type} interface`}
            title={`${type} interface`}
          />
          )
        }
        {
          !getTimelineImage(type)
          && `${type} Interface`
        }
      </div>
    </motion.div>
    <div className="timeline-stage__meta">
      <h2 className="timeline-stage__title">{label || '\u00A0'}</h2>
      <div className="timeline-stage__icons">
        { hasFilter
          && (
          <div className="timeline-stage__icon">
            <img src={filterIcon} alt="Filter" title="Filter" />
          </div>
          )}
        { hasSkipLogic
          && (
          <div className="timeline-stage__icon">
            <img src={skipLogicIcon} alt="Skip logic" title="Skip logic" />
          </div>
          )}
      </div>
    </div>
  </div>
));

EditStageButton.propTypes = {
  hasFilter: PropTypes.bool,
  hasSkipLogic: PropTypes.bool,
  label: PropTypes.string,
  layoutId: PropTypes.string,
  onEditStage: PropTypes.func.isRequired,
  type: PropTypes.string.isRequired,
};

EditStageButton.defaultProps = {
  hasFilter: false,
  hasSkipLogic: false,
  label: null,
  layoutId: null,
};

export default EditStageButton;
