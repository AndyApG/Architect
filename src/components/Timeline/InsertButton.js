import React from 'react';
import { motion } from 'framer-motion';
import PropTypes from 'prop-types';

const InsertButton = ({ onClick }) => (
  <motion.div
    className="timeline__insert"
    onClick={onClick}
    // initial={{ opacity: 0 }}
    // animate={{ opacity: 1 }}
    // exit={{ opacity: 0 }}
  >
    Add stage here
  </motion.div>
);

InsertButton.propTypes = {
  onClick: PropTypes.func.isRequired,
};

export default InsertButton;
