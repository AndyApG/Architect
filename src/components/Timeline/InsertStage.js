import React from 'react';

const InsertStage = () => (
  <div className="timeline-insert-stage">
    insert stage<br />
    insert stage<br />
    insert stage<br />
    insert stage<br />
    insert stage<br />
    insert stage<br />
    insert stage<br />
  </div>
);

export default InsertStage;
