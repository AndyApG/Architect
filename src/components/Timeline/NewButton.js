import React from 'react';
import { Icon } from '@codaco/ui';

const NewButton = props => (
  <div className="list-new-button" {...props}>
    <Icon name="add" />
  </div>
);

export { NewButton };

export default NewButton;
