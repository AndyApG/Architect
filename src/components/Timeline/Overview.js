import React from 'react';

const TimelineOverview = () => (
  <div className="timeline-overview">
    <div className="panel">
      <br />
      <br />
      <br />
    </div>
  </div>
);

export default TimelineOverview;
