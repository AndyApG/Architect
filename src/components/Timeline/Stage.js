import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';
import cx from 'classnames';
import { Zoom } from '../../behaviours';
import { getCSSVariableAsString } from '../../utils/CSSVariables';
import timelineImages from '../../images/timeline';

const zoomColors = [getCSSVariableAsString('--light-background'), '#ffffff'];

const getTimelineImage = type =>
  get(timelineImages, type);

const EditStageButton = Zoom(
  ({
    onEditStage,
    type,
    label,
  }) => (
    <div className="timeline-stage__edit-stage">
      <div className="timeline-stage__edit-stage-title">{label || '\u00A0'}</div>
      <div
        className="timeline-stage__screen"
        role="button"
        onClick={onEditStage}
        tabIndex="0"
      >
        <div className="timeline-stage__screen-preview">
          {
            getTimelineImage(type) &&
            <img
              src={getTimelineImage(type)}
              alt={`${type} Interface`}
              title={`${type} Interface`}
            />
          }
          {
            !getTimelineImage(type) &&
            `${type} Interface`
          }
        </div>
      </div>
    </div>

  ),
);

const ConfigureSkipLogicButton = Zoom(
  ({
    onEditSkipLogic,
  }) => (
    <a
      className="timeline-stage__control"
      onClick={onEditSkipLogic}
    >
      <div className="timeline-stage__control-icon">&#9881;</div>
      Configure skip logic
    </a>
  ),
);

class TimelineStage extends PureComponent {
  static propTypes = {
    className: PropTypes.string,
    label: PropTypes.string,
    type: PropTypes.string.isRequired,
    onEditStage: PropTypes.func.isRequired,
    onDeleteStage: PropTypes.func.isRequired,
    onEditSkipLogic: PropTypes.func.isRequired,
    onInsertStage: PropTypes.func.isRequired,
  };

  static defaultProps = {
    label: '',
    className: '',
  };

  render() {
    const {
      className,
      onEditStage,
      onDeleteStage,
      onEditSkipLogic,
      onInsertStage,
      type,
      label,
      ...rest
    } = this.props;

    const componentClasses = cx(
      'timeline-stage',
      className,
    );

    return (
      <div className={componentClasses} {...rest}>
        <div
          className="timeline-stage__notch"
          onClick={onEditStage}
        />
        <EditStageButton
          onEditStage={onEditStage}
          type={type}
          label={label}
          zoomColors={zoomColors}
        />
        <div className="timeline-stage__controls">
          <a
            className="timeline-stage__control"
            onClick={() => onInsertStage(0)}
          >
            <div className="timeline-stage__control-icon">↑</div>
            Add screen before
          </a>
          <ConfigureSkipLogicButton
            onEditSkipLogic={onEditSkipLogic}
            zoomColors={['#ffffff', '#f2b700']}
          />
          <a
            className="timeline-stage__control"
            onClick={onDeleteStage}
          >
            <div className="timeline-stage__control-icon">✕</div>
            Delete stage
          </a>
          <a
            className="timeline-stage__control"
            onClick={() => onInsertStage(1)}
          >
            <div className="timeline-stage__control-icon">↓</div>
            Add screen after
          </a>
        </div>
      </div>
    );
  }
}

export default TimelineStage;
