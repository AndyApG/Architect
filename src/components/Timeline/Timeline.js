import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { TransitionGroup } from 'react-transition-group';
import { get } from 'lodash';
import { Drawer } from '../Transitions';
import Overview from './Overview';
import Stage from './Stage';
import AddNew from './AddNew';
import InsertStage from './InsertStage';
import constrain from '../../behaviours/constrain';

class Timeline extends PureComponent {
  static propTypes = {
    stages: PropTypes.array,
    onCreateStage: PropTypes.func,
    onEditStage: PropTypes.func,
    onEditSkipLogic: PropTypes.func,
  };

  static defaultProps = {
    stages: [],
    onCreateStage: () => {},
    onEditStage: () => {},
    onEditSkipLogic: () => {},
  };

  constructor(props) {
    super(props);

    this.state = {
      highlightY: 0,
      highlightHide: false,
      insertStageAtIndex: null,
    };
  }

  onMouseEnterStage = (e) => {
    const offset = e.target.closest('.timeline-stage').offsetTop;
    this.setState({ highlightY: offset, highlightHide: false });
  };

  onMouseLeaveStage = () => {
    this.setState({ highlightHide: true });
  };

  onInsertStage = (index) => {
    this.setState({ insertStageAtIndex: index, highlightHide: true });
  };

  createStage = (type, index) => {
    this.setState({ insertStageAtIndex: null, highlightHide: true });
    this.props.onCreateStage(type, index);
  };

  hasStages = () => this.props.stages.length > 0;

  injectInsertStage = (items) => {
    const { insertStageAtIndex } = this.state;

    if (insertStageAtIndex !== null) {
      return [
        ...items.slice(0, insertStageAtIndex),
        this.renderInsertStage(insertStageAtIndex),
        ...items.slice(insertStageAtIndex),
      ];
    }

    return items;
  }

  renderInsertStage = insertStageAtIndex => (
    <Drawer
      key={`insert-${insertStageAtIndex}`}
      timeout={1000}
      unmountOnExit
    >
      <InsertStage
        onSelectStageType={type => this.createStage(type, insertStageAtIndex)}
      />
    </Drawer>
  );

  renderHighlight = () => {
    const opacity = this.state.highlightHide ? 0 : 1;
    const transform = this.state.highlightHide ? 'translateY(0px)' : `translateY(${this.state.highlightY}px)`;

    const highlightStyles = {
      transform,
      opacity,
    };

    return (
      <div
        className="timeline__stages-highlight"
        key="highlight"
        style={highlightStyles}
      />
    );
  }

  renderStages = () => {
    const items = this.props.stages.map(this.renderStage);

    return this.injectInsertStage(items);
  }

  renderStage = (stage, index) => (
    <Stage
      key={`stage_${stage.id}`}
      id={stage.id}
      type={stage.type}
      label={stage.label}
      onMouseEnter={this.onMouseEnterStage}
      onMouseLeave={this.onMouseLeaveStage}
      onEditStage={() => this.props.onEditStage(stage.id)}
      onInsertStage={position => this.onInsertStage(index + position)}
      onEditSkipLogic={() => this.props.onEditSkipLogic(stage.id)}
    />
  );

  render() {
    return (
      <div className="timeline">
        <div className="timeline__main">
          <div className="timeline__background" />
          <div className="timeline__content">
            <Overview
              title="My protocol"
            />
            <div className="timeline__stages">
              { this.hasStages() && this.renderHighlight() }
              <TransitionGroup component={null}>
                { this.renderStages() }
              </TransitionGroup>
              { !this.hasStages() && <AddNew onInsertStage={() => this.props.onInsertStage(0)} /> }
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export { Timeline };

export default constrain([60, 0, 0, 0])(Timeline);
