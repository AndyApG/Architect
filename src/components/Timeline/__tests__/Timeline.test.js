/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { Timeline } from '../Timeline';

const mockProps = {
  deleteStage: () => {},
  goTo: () => {},
  openDialog: () => {},
};

describe('<Timeline />', () => {
  it('can render', () => {
    const component = shallow(<Timeline {...mockProps} />);

    expect(component).toMatchSnapshot();
  });

  it('renders stages', () => {
    const mockStages = [{ id: 1, type: 'NameGenerator' }, { id: 2, type: 'Sociogram' }];

    const component = shallow(<Timeline {...mockProps} stages={mockStages} />);

    expect(component.find('sortableElement(Stage)').length).toEqual(2);
  });
});
