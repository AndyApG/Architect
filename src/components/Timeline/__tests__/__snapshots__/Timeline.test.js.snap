// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timeline /> can render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Timeline
    deleteStage={[Function]}
    goTo={[Function]}
    overview={Object {}}
    show={true}
    sorting={false}
    stages={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="timeline__stages"
      >
        <TransitionGroup
          childFactory={[Function]}
          component="div"
        />
        <InsertStage
          onCancel={null}
          onSelectStage={[Function]}
        />
      </div>,
      "className": "timeline timeline--show",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          <TransitionGroup
            childFactory={[Function]}
            component="div"
          />,
          <InsertStage
            onCancel={null}
            onSelectStage={[Function]}
          />,
        ],
        "className": "timeline__stages",
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "childFactory": [Function],
            "children": Array [],
            "component": "div",
          },
          "ref": null,
          "rendered": Array [],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "insertStage",
          "nodeType": "class",
          "props": Object {
            "onCancel": null,
            "onSelectStage": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="timeline__stages"
        >
          <TransitionGroup
            childFactory={[Function]}
            component="div"
          />
          <InsertStage
            onCancel={null}
            onSelectStage={[Function]}
          />
        </div>,
        "className": "timeline timeline--show",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            false,
            <TransitionGroup
              childFactory={[Function]}
              component="div"
            />,
            <InsertStage
              onCancel={null}
              onSelectStage={[Function]}
            />,
          ],
          "className": "timeline__stages",
        },
        "ref": null,
        "rendered": Array [
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "childFactory": [Function],
              "children": Array [],
              "component": "div",
            },
            "ref": null,
            "rendered": Array [],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "insertStage",
            "nodeType": "class",
            "props": Object {
              "onCancel": null,
              "onSelectStage": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
